cloudflare_record does not handle deletion outside of Terraform

### Confirmation

- [X] My issue isn't already found on the issue tracker.
- [X] I have replicated my issue using the latest version of the provider and it is still present.

### Terraform and Cloudflare provider version

```
Terraform v1.1.7
on linux_amd64
+ provider registry.terraform.io/cloudflare/cloudflare v3.14.0

Your version of Terraform is out of date! The latest version
is 1.1.9. You can update by downloading from https://www.terraform.io/downloads.html
```

### Affected resource(s)

* cloudflare_record

Judging from the issue history, this probably affects other resources too.

### Terraform configuration files

```hcl
terraform {
  required_providers {
    cloudflare = {
      source = "cloudflare/cloudflare"
      version = ">= 3.14.0"
    }
  }
}

provider "cloudflare" {
  account_id = chomp(file("./account_id"))
}

resource "cloudflare_record" "myhost" {
  zone_id = chomp(file("./zone_id"))
  name = "myhost"
  type = "A"
  value = "12.34.56.78"
  proxied = false
}
```


### Debug output

```
2022-05-12T06:00:48.418-0500 [INFO]  Terraform version: 1.1.7
2022-05-12T06:00:48.418-0500 [INFO]  Go runtime version: go1.17.8
2022-05-12T06:00:48.418-0500 [INFO]  CLI args: []string{"terraform", "apply", "-no-color"}
2022-05-12T06:00:48.418-0500 [DEBUG] Attempting to open CLI config file: /home/kdavis/.terraformrc
2022-05-12T06:00:48.419-0500 [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
2022-05-12T06:00:48.419-0500 [DEBUG] ignoring non-existing provider search directory terraform.d/plugins
2022-05-12T06:00:48.419-0500 [DEBUG] ignoring non-existing provider search directory /home/kdavis/.terraform.d/plugins
2022-05-12T06:00:48.419-0500 [DEBUG] ignoring non-existing provider search directory /home/kdavis/.local/share/terraform/plugins
2022-05-12T06:00:48.419-0500 [DEBUG] ignoring non-existing provider search directory /usr/local/share/terraform/plugins
2022-05-12T06:00:48.419-0500 [DEBUG] ignoring non-existing provider search directory /usr/share/terraform/plugins
2022-05-12T06:00:48.419-0500 [INFO]  CLI command args: []string{"apply", "-no-color"}
2022-05-12T06:00:48.420-0500 [DEBUG] New state was assigned lineage "76aa3e91-9d09-0a28-a923-106afc2ac730"
2022-05-12T06:00:48.482-0500 [DEBUG] checking for provisioner in "."
2022-05-12T06:00:48.489-0500 [DEBUG] checking for provisioner in "/usr/bin"
2022-05-12T06:00:48.490-0500 [INFO]  backend/local: starting Apply operation
2022-05-12T06:00:48.492-0500 [DEBUG] created provider logger: level=debug
2022-05-12T06:00:48.492-0500 [INFO]  provider: configuring client automatic mTLS
2022-05-12T06:00:48.501-0500 [DEBUG] provider: starting plugin: path=.terraform/providers/registry.terraform.io/cloudflare/cloudflare/3.14.0/linux_amd64/terraform-provider-cloudflare_v3.14.0 args=[.terraform/providers/registry.terraform.io/cloudflare/cloudflare/3.14.0/linux_amd64/terraform-provider-cloudflare_v3.14.0]
2022-05-12T06:00:48.502-0500 [DEBUG] provider: plugin started: path=.terraform/providers/registry.terraform.io/cloudflare/cloudflare/3.14.0/linux_amd64/terraform-provider-cloudflare_v3.14.0 pid=22753
2022-05-12T06:00:48.502-0500 [DEBUG] provider: waiting for RPC address: path=.terraform/providers/registry.terraform.io/cloudflare/cloudflare/3.14.0/linux_amd64/terraform-provider-cloudflare_v3.14.0
2022-05-12T06:00:48.507-0500 [INFO]  provider.terraform-provider-cloudflare_v3.14.0: configuring server automatic mTLS: timestamp=2022-05-12T06:00:48.507-0500
2022-05-12T06:00:48.526-0500 [DEBUG] provider.terraform-provider-cloudflare_v3.14.0: plugin address: address=/tmp/plugin216648483 network=unix timestamp=2022-05-12T06:00:48.525-0500
2022-05-12T06:00:48.526-0500 [DEBUG] provider: using plugin: version=5
2022-05-12T06:00:48.576-0500 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = transport is closing"
2022-05-12T06:00:48.576-0500 [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.terraform.io/cloudflare/cloudflare/3.14.0/linux_amd64/terraform-provider-cloudflare_v3.14.0 pid=22753
2022-05-12T06:00:48.576-0500 [DEBUG] provider: plugin exited
2022-05-12T06:00:48.577-0500 [DEBUG] Building and walking validate graph
2022-05-12T06:00:48.577-0500 [DEBUG] ProviderTransformer: "cloudflare_record.myhost" (*terraform.NodeValidatableResource) needs provider["registry.terraform.io/cloudflare/cloudflare"]
2022-05-12T06:00:48.577-0500 [DEBUG] ReferenceTransformer: "cloudflare_record.myhost" references: []
2022-05-12T06:00:48.577-0500 [DEBUG] ReferenceTransformer: "provider[\"registry.terraform.io/cloudflare/cloudflare\"]" references: []
2022-05-12T06:00:48.578-0500 [DEBUG] Starting graph walk: walkValidate
2022-05-12T06:00:48.578-0500 [DEBUG] created provider logger: level=debug
2022-05-12T06:00:48.578-0500 [INFO]  provider: configuring client automatic mTLS
2022-05-12T06:00:48.587-0500 [DEBUG] provider: starting plugin: path=.terraform/providers/registry.terraform.io/cloudflare/cloudflare/3.14.0/linux_amd64/terraform-provider-cloudflare_v3.14.0 args=[.terraform/providers/registry.terraform.io/cloudflare/cloudflare/3.14.0/linux_amd64/terraform-provider-cloudflare_v3.14.0]
2022-05-12T06:00:48.588-0500 [DEBUG] provider: plugin started: path=.terraform/providers/registry.terraform.io/cloudflare/cloudflare/3.14.0/linux_amd64/terraform-provider-cloudflare_v3.14.0 pid=22762
2022-05-12T06:00:48.588-0500 [DEBUG] provider: waiting for RPC address: path=.terraform/providers/registry.terraform.io/cloudflare/cloudflare/3.14.0/linux_amd64/terraform-provider-cloudflare_v3.14.0
2022-05-12T06:00:48.593-0500 [INFO]  provider.terraform-provider-cloudflare_v3.14.0: configuring server automatic mTLS: timestamp=2022-05-12T06:00:48.593-0500
2022-05-12T06:00:48.611-0500 [DEBUG] provider.terraform-provider-cloudflare_v3.14.0: plugin address: address=/tmp/plugin2325534775 network=unix timestamp=2022-05-12T06:00:48.611-0500
2022-05-12T06:00:48.611-0500 [DEBUG] provider: using plugin: version=5
2022-05-12T06:00:48.670-0500 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = transport is closing"
2022-05-12T06:00:48.671-0500 [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.terraform.io/cloudflare/cloudflare/3.14.0/linux_amd64/terraform-provider-cloudflare_v3.14.0 pid=22762
2022-05-12T06:00:48.671-0500 [DEBUG] provider: plugin exited
2022-05-12T06:00:48.672-0500 [INFO]  backend/local: apply calling Plan
2022-05-12T06:00:48.672-0500 [DEBUG] Building and walking plan graph for NormalMode
2022-05-12T06:00:48.672-0500 [DEBUG] ProviderTransformer: "cloudflare_record.myhost (expand)" (*terraform.nodeExpandPlannableResource) needs provider["registry.terraform.io/cloudflare/cloudflare"]
2022-05-12T06:00:48.672-0500 [DEBUG] ReferenceTransformer: "cloudflare_record.myhost (expand)" references: []
2022-05-12T06:00:48.672-0500 [DEBUG] ReferenceTransformer: "provider[\"registry.terraform.io/cloudflare/cloudflare\"]" references: []
2022-05-12T06:00:48.672-0500 [DEBUG] Starting graph walk: walkPlan
2022-05-12T06:00:48.672-0500 [DEBUG] created provider logger: level=debug
2022-05-12T06:00:48.672-0500 [INFO]  provider: configuring client automatic mTLS
2022-05-12T06:00:48.681-0500 [DEBUG] provider: starting plugin: path=.terraform/providers/registry.terraform.io/cloudflare/cloudflare/3.14.0/linux_amd64/terraform-provider-cloudflare_v3.14.0 args=[.terraform/providers/registry.terraform.io/cloudflare/cloudflare/3.14.0/linux_amd64/terraform-provider-cloudflare_v3.14.0]
2022-05-12T06:00:48.681-0500 [DEBUG] provider: plugin started: path=.terraform/providers/registry.terraform.io/cloudflare/cloudflare/3.14.0/linux_amd64/terraform-provider-cloudflare_v3.14.0 pid=22770
2022-05-12T06:00:48.681-0500 [DEBUG] provider: waiting for RPC address: path=.terraform/providers/registry.terraform.io/cloudflare/cloudflare/3.14.0/linux_amd64/terraform-provider-cloudflare_v3.14.0
2022-05-12T06:00:48.687-0500 [INFO]  provider.terraform-provider-cloudflare_v3.14.0: configuring server automatic mTLS: timestamp=2022-05-12T06:00:48.686-0500
2022-05-12T06:00:48.705-0500 [DEBUG] provider.terraform-provider-cloudflare_v3.14.0: plugin address: address=/tmp/plugin2849232429 network=unix timestamp=2022-05-12T06:00:48.705-0500
2022-05-12T06:00:48.705-0500 [DEBUG] provider: using plugin: version=5
2022-05-12T06:00:48.766-0500 [WARN]  ValidateProviderConfig from "provider[\"registry.terraform.io/cloudflare/cloudflare\"]" changed the config value, but that value is unused
2022-05-12T06:00:48.767-0500 [INFO]  provider.terraform-provider-cloudflare_v3.14.0: 2022/05/12 06:00:48 [INFO] Cloudflare Client configured for user:: timestamp=2022-05-12T06:00:48.767-0500
2022-05-12T06:00:48.767-0500 [INFO]  provider.terraform-provider-cloudflare_v3.14.0: 2022/05/12 06:00:48 [INFO] Using specified account id REDACTED in Cloudflare provider: timestamp=2022-05-12T06:00:48.767-0500
2022-05-12T06:00:48.767-0500 [INFO]  provider.terraform-provider-cloudflare_v3.14.0: 2022/05/12 06:00:48 [INFO] Cloudflare Client configured for user:: timestamp=2022-05-12T06:00:48.767-0500
2022-05-12T06:00:48.767-0500 [DEBUG] ReferenceTransformer: "cloudflare_record.myhost" references: []
cloudflare_record.myhost: Refreshing state... [id=REDACTED]
2022-05-12T06:00:48.771-0500 [INFO]  provider.terraform-provider-cloudflare_v3.14.0: 2022/05/12 06:00:48 [DEBUG] Cloudflare API Request Details:
---[ REQUEST ]---------------------------------------
GET /client/v4/zones/REDACTED/dns_records/REDACTED HTTP/1.1
Host: api.cloudflare.com
User-Agent: terraform/1.1.7 terraform-plugin-sdk/2.10.1 terraform-provider-cloudflare/3.14.0
Authorization: Bearer REDACTED
Content-Type: application/json
Accept-Encoding: gzip


-----------------------------------------------------: timestamp=2022-05-12T06:00:48.771-0500
2022-05-12T06:00:49.633-0500 [INFO]  provider.terraform-provider-cloudflare_v3.14.0: 2022/05/12 06:00:49 [DEBUG] Cloudflare API Response Details:
---[ RESPONSE ]--------------------------------------
HTTP/1.1 404 Not Found
Connection: close
Transfer-Encoding: chunked
Cf-Cache-Status: DYNAMIC
Cf-Ray: REDACTED
Content-Type: application/json
Date: Thu, 12 May 2022 11:00:49 GMT
Expect-Ct: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
Server: cloudflare
Set-Cookie: __cflb=REDACTED; SameSite=Lax; path=/; expires=Thu, 12-May-22 13:30:50 GMT; HttpOnly
Set-Cookie: __cfruid=REDACTED; path=/; domain=.api.cloudflare.com; HttpOnly; Secure; SameSite=None
Vary: Accept-Encoding
X-Envoy-Upstream-Service-Time: 13

6a
{
 "result": null,
 "success": false,
 "errors": [
  {
   "code": 81044,
   "message": "Record does not exist."
  }
 ],
 "messages": []
}
0


-----------------------------------------------------: timestamp=2022-05-12T06:00:49.632-0500
2022-05-12T06:00:49.634-0500 [ERROR] vertex "cloudflare_record.myhost" error: Record does not exist. (81044)
2022-05-12T06:00:49.634-0500 [ERROR] vertex "cloudflare_record.myhost" error: Record does not exist. (81044)
2022-05-12T06:00:49.634-0500 [ERROR] vertex "cloudflare_record.myhost (expand)" error: Record does not exist. (81044)

Error: Record does not exist. (81044)

  with cloudflare_record.myhost,
  on main.tf line 14, in resource "cloudflare_record" "myhost":
  14: resource "cloudflare_record" "myhost" {

2022-05-12T06:00:49.635-0500 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = transport is closing"
2022-05-12T06:00:49.637-0500 [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.terraform.io/cloudflare/cloudflare/3.14.0/linux_amd64/terraform-provider-cloudflare_v3.14.0 pid=22770
2022-05-12T06:00:49.637-0500 [DEBUG] provider: plugin exited
```

### Panic output

_No response_

### Expected output

`terraform apply` reports something along the lines of:

> Detected changes outside of Terraform: cloudflare_record.myhost deleted
>
> Terraform will perform the following actions: create cloudflare_record.myhost 

### Actual output

`terraform apply` exits with an error:
```
cloudflare_record.myhost: Refreshing state... [id=REDACTED]
╷
│ Error: Record does not exist. (81044)
│
│   with cloudflare_record.myhost,
│   on main.tf line 14, in resource "cloudflare_record" "myhost":
│   14: resource "cloudflare_record" "myhost" {
│
```

No attempt is made to create cloudflare_record.myhost again.

### Steps to reproduce

* Create the resource using `terraform apply`
* Delete the resource through the Cloudflare console
* Run `terraform apply` again

### Additional factoids

_No response_

### References

* Same issue for cloudflare_healthcheck: #785
* Same issue for cloudflare_ip_list: #794

Please could this be reviewed and fixed for all of this provider's resources?

--- a/.changelog/1740.txt
+++ b/.changelog/1740.txt
@@ -1,0 +1,3 @@
+```release-note:enhancement
+resource/cloudflare_record: Validate that DNS record names are non-empty
+```

--- a/internal/provider/resource_cloudflare_record.go
+++ b/internal/provider/resource_cloudflare_record.go
@@ -99,8 +99,8 @@ func resourceCloudflareRecordCreate(ctx context.Context, d *schema.ResourceData,
 	}
 
 	// Validate value based on type
-	if err := validateRecordName(newRecord.Type, newRecord.Content); err != nil {
-		return diag.FromErr(fmt.Errorf("error validating record name %q: %w", newRecord.Name, err))
+	if err := validateRecordContent(newRecord.Type, newRecord.Content); err != nil {
+		return diag.FromErr(fmt.Errorf("error validating record content of %q: %w", newRecord.Name, err))
 	}
 
 	var proxiedVal *bool

--- a/internal/provider/resource_cloudflare_record.go
+++ b/internal/provider/resource_cloudflare_record.go
@@ -98,6 +98,10 @@ func resourceCloudflareRecordCreate(ctx context.Context, d *schema.ResourceData,
 		newRecord.TTL = ttl.(int)
 	}
 
+	if newRecord.Name == "" {
+		return diag.FromErr(fmt.Errorf("record on zone %s must not have an empty name (use @ for the zone apex)", newRecord.ZoneID))
+	}
+
 	// Validate value based on type
 	if err := validateRecordContent(newRecord.Type, newRecord.Content); err != nil {
 		return diag.FromErr(fmt.Errorf("error validating record content of %q: %w", newRecord.Name, err))

--- a/internal/provider/resource_cloudflare_record.go
+++ b/internal/provider/resource_cloudflare_record.go
@@ -189,8 +189,8 @@ func resourceCloudflareRecordRead(ctx context.Context, d *schema.ResourceData, m
 
 	record, err := client.DNSRecord(ctx, zoneID, d.Id())
 	if err != nil {
-		if strings.Contains(err.Error(), "Invalid dns record identifier") ||
-			strings.Contains(err.Error(), "HTTP status 404") {
+		var notFoundError *cloudflare.NotFoundError
+		if errors.As(err, &notFoundError) {
 			tflog.Warn(ctx, fmt.Sprintf("Removing record from state because it's not found in API"))
 			d.SetId("")
 			return nil

--- a/internal/provider/resource_cloudflare_record_test.go
+++ b/internal/provider/resource_cloudflare_record_test.go
@@ -476,7 +476,7 @@ func TestAccCloudflareRecord_MXWithPriorityZero(t *testing.T) {
 				Config: testAccCheckCloudflareRecordConfigMXWithPriorityZero(zoneID, rnd, zoneName),
 				Check: resource.ComposeTestCheckFunc(
 					resource.TestCheckResourceAttr(resourceName, "priority", "0"),
-					resource.TestCheckResourceAttr(resourceName, "value", "."),
+					resource.TestCheckResourceAttr(resourceName, "value", "mail.terraform.cfapi.net"),
 				),
 			},
 		},
@@ -772,7 +772,7 @@ func testAccCheckCloudflareRecordConfigMXWithPriorityZero(zoneID, name, zoneName
 resource "cloudflare_record" "%[2]s" {
 	zone_id = "%[1]s"
 	name = "%[2]s"
-	value = "."
+	value = "mail.terraform.cfapi.net"
 	type = "MX"
 	priority = 0
 	proxied = false

--- a/internal/provider/validators.go
+++ b/internal/provider/validators.go
@@ -27,9 +27,9 @@ func validateRecordType(t string, proxied bool) error {
 	return fmt.Errorf("type %q cannot be proxied", t)
 }
 
-// validateRecordName ensures that based on supplied record type, the name content matches
-// Currently only validates A and AAAA types.
-func validateRecordName(t string, value string) error {
+// validateRecordContent ensures that the record's content is valid for the
+// supplied record type. Currently only validates A and AAAA types.
+func validateRecordContent(t string, value string) error {
 	switch t {
 	case "A":
 		// Must be ipv4 addr

--- a/internal/provider/validators_test.go
+++ b/internal/provider/validators_test.go
@@ -51,8 +51,8 @@ func TestValidateRecordName(t *testing.T) {
 	}
 
 	for k, v := range validNames {
-		if err := validateRecordName(k, v); err != nil {
-			t.Fatalf("%q should be a valid name for type %q: %v", v, k, err)
+		if err := validateRecordContent(k, v); err != nil {
+			t.Fatalf("%q should be valid content for type %q: %v", v, k, err)
 		}
 	}
 
@@ -62,8 +62,8 @@ func TestValidateRecordName(t *testing.T) {
 		"TXT":  "\n",
 	}
 	for k, v := range invalidNames {
-		if err := validateRecordName(k, v); err == nil {
-			t.Fatalf("%q should be an invalid name for type %q", v, k)
+		if err := validateRecordContent(k, v); err == nil {
+			t.Fatalf("%q should be invalid content for type %q", v, k)
 		}
 	}
 }

